### PR TITLE
[REVIEW] Remove utility.hpp in favor of error.hpp w/libcudf style

### LINKED ISSUE #153: [BUG] Align utility.hpp with libcudf style
0. Move all error handling utilities to a new header called error.hpp
1. Add a `CUSPATIAL_EXPECTS` to use in place of `CUDF_EXPECTS`. Add appropriate cuspatial exception types, e.g. `cuspatial::logic_error`.
2.  `HANDLE_CUSPARSE_STATUS` should become `CUSPARSE_TRY`. It should use `CUSPATIAL_EXPECTS` and have no raw `printf`.
3. `HANDLE_CUDA_ERROR` should become `CUDA_TRY`, identical to the one in libcudf. It should throw, not `exit()`!

### PR DESCRIPTION
**Pull request by cwharris**
closes #153

### CODE REVIEW

File: cpp/include/cuspatial/error.hpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cusparse.h>
```
Comment (harrism) [inline cpp/include/cuspatial/error.hpp:19]: This causes every cuspatial file to include a ton of code it doesn't use.  Instead, just forward declare `cuSparseStatus_t`. Better yet, move all the cusparse stuff to a separate utility header.

Comment (cwharris) [reply cpp/include/cuspatial/error.hpp:19]: When you say move to a separate header, do you mean forward-declare everything we need from `cusparse.h`, or just `cuSparseStatus_t`?

Comment (harrism) [reply cpp/include/cuspatial/error.hpp:19]: I think if you make a separate cusparse_error.hpp you can include cusparse.h from it and you don't have to forward decl anything.  This just saves all of cuspatial having to include cusparse.h except for the one file where we use it.

**Comment by cwharris** [comment]
rerun tests

**Comment by harrism** [comment]
Still getting killed by these errors on gcc 7.x: 

```
/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h(1761): error: identifier "__builtin_ia32_sqrtsd_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h(1770): error: identifier "__builtin_ia32_sqrtss_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h(2728): error: identifier "__builtin_ia32_scalefsd_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h(2737): error: identifier "__builtin_ia32_scalefss_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h(11265): error: identifier "__builtin_ia32_scalefsd_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512fintrin.h(11274): error: identifier "__builtin_ia32_scalefss_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512dqintrin.h(1163): error: identifier "__builtin_ia32_reducesd" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512dqintrin.h(1171): error: identifier "__builtin_ia32_reducess" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512dqintrin.h(1179): error: identifier "__builtin_ia32_rangesd128_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512dqintrin.h(1189): error: identifier "__builtin_ia32_rangess128_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512dqintrin.h(1198): error: identifier "__builtin_ia32_rangesd128_round" is undefined

/usr/lib/gcc/x86_64-linux-gnu/7/include/avx512dqintrin.h(1207): error: identifier "__builtin_ia32_rangess128_round" is undefined
```

Not sure what to do...

**Comment by harrism** [comment]
#169 fixes the CI  failure.

**Pull request merged**